File: 15-3sum/15-3sum.py
# Approach: Sorting + Two Pointers
# Time Complexity: O(n2)
# Space Complexity: O(logn) to O(n)
    
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        length = len(nums)
        if length < 3:
            return []
    
        nums.sort()
        if nums[0] > 0 or nums[length - 1] < 0:
            return []
        
        res = []
        
        for i in range(length):
            if nums[i] > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            l, r = i + 1, length - 1
            while l < r:
                sum = nums[i] + nums[l] + nums[r]
                if sum < 0:
                    l += 1
                elif sum > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l< r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1
        return res                 
                        
        
        
        
        

File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  if (nums.length < 3) {
    return [];
  }
  // Sort the input array nums
  const arr = nums.sort((a,b) => a-b);
  
  if (arr[0] > 0 || arr[arr.length - 1] < 0) {
    return [];
  }
    
  const n = arr.length;
  const res = [];
  for (let i = 0; i < n; i ++) {
    // If the current value is greater than zero, break from the loop
    if (nums[i] > 0) {
      return res;
    }
    // If the current value is the same as the one before, skip it
    if (i > 0 && arr[i] === arr[i - 1]) {
      continue;
    }
    // Otherwise two sum using two pointers for the current position i
    let l = i + 1;
    let r = n - 1;
    while(l < r) {
      const temp = arr[i] + arr[l] + arr[r];
      if (temp > 0) {
        r --;
      }
      if (temp < 0) {
        l ++;
      }
      if (temp === 0) {
        res.push([nums[i], nums[l], nums[r]]);
        // skip dup
        while(l < r && nums[l] === nums[l + 1]) {
          l ++;
        }
        while(l < r && nums[r] === nums[r - 1]) {
          r --;
        }
        l ++;
        r --;
      }
    }
  }
  return res;
};